### PR TITLE
Design 2: MiniCluster concept with Manager

### PR DESCRIPTION
**Pull request by vsoch**
I had fun this weekend! This is a better start to a design, and I've actually gotten us back to where we were on the current main branch (e.g., figuring out the right way to launch the job). I don't have an answer to this yet - I got close because I figured out how to exec a command to a pod, but there is a permissions error preventing it from working. We will likely still need to figure out the etc-hosts (or creating a network config for it). I can put ideas in a separate issue, but I've found [host aliases](https://kubernetes.io/docs/tasks/network/customize-hosts-file-for-pods/), or we could [watch pods](https://github.com/kubeflow/mpi-operator/blob/d7fc50603a4d6963f18844a4145ea33ab632fe93/pkg/controllers/kubectl_delivery/controller.go#L82) and retrieve the host ip addresses post creation, and populate /etc/hosts e.g., [here](https://github.com/kubeflow/mpi-operator/blob/d7fc50603a4d6963f18844a4145ea33ab632fe93/pkg/controllers/kubectl_delivery/controller.go#L142-L159). If we can have some confidence this hack won't lead to bad outcomes (pods being re-created many times and this generation being triggered so many times everything gets stuck) it could work! It seems fun to try anyway... :laughing: 

Ping @zekemorton @milroy @ArangoGutierrez !

design-two-v

### CODE REVIEW

**Comment by milroy** [comment]
@vsoch this is really great progress; thank you for the organization and diagram! 

I have a few high-level questions to prime my understanding (enumerated by the items in the diagram):

2. How will Flux Setup update the cluster status? Which K8s cluster objects is it updating?
3. Is there any interaction between the Waiting queue and the K8s default scheduler? (It appears not, but checking.)
5/6a. Are the `StatefulSet` Pods created at this stage as components of the `MiniCluster` by the `FluxJob` reconciler? What determines when the resources are available (i.e. K8s default scheduler or KubeFlux)? What criteria allow the job to enter the queue heap (is allowed to run)?
5/6b.  "Moving from waiting -> the manager->queue->heap indicates running, and this is done by the scheduler." This is the Flux Operator scheduler, not the K8s cluster scheduler (e.g., the K8s default), correct?
5/6c. Under what conditions does the `MiniCluster` transition to READY (`StatefulSet` Pods + `ConfigMap` + `Secrets` are READY)?

And unrelated to the enumerated sequence, can the current design work with `PodGroup`s?

**Comment by vsoch** [comment]
Here are some quick answers (and is there a meeting we can chat about more details this week?)

> 2. How will Flux Setup update the cluster status? Which K8s cluster objects is it updating?

The FluxSetup started as a concept to hold asset defaults like a broker config and /etc/hosts, but soon it became apparent that 1. these components could all be generated automatically, and 2. they fell under some concept of a MiniCluster (at the time TBA for how that design would work). There was also a conflict that the FluxJob (what the user submits) had a container that would need to be the same as the one in the stateful set (at least per the diagram on Confluence). I decided instead that I'd call the set of components a MiniCluster, and it should be controlled by a FluxJob because largely the user is dictating what they want and need.

This leaves us with the FluxSetup. I think the FluxSetup should largely be managing the general setup for the cluster - resources available / max sizes and things like that. This goes hand and hand with the Scheduler (controller?) would could run as a piece of FluxSetup, but I decided to implement it separately since it's just a loop to (eventually check state) and then move jobs (wanting mini clusters) from waiting to running. This isn't implemented yet, but I'm hoping a FluxSetup can simply watch for the creation / deletion of MiniClusters by the FluxJob (assuming they either clean themselves up or are cleaned up per request by the user, maybe this could be a variable?) or given a change by the admin to the setup (e.g., "Scale this down to X") it can move some number of jobs from running to waiting, and then the FluxJob will detect this and respond appropriately. So to more directly answer your question, at least my current thinking is that FluxSetup is more observing the cluster and interacting with the fluxManager (shared by reconcilers) to dictate outcomes, and (currently) it's not updating any objects directly, but it's watching FluxJob. If we could find a more elegant way for FluxSetup to be able to manage the MiniCluster and then FluxJob to just use it, that would work too. I think the key component I'm missing is how exactly we are running the jobs once we have the cluster - once we figure that out I think we can try different controller designs. 

> 3. Is there any interaction between the Waiting queue and the K8s default scheduler? (It appears not, but checking.)

I don't think so! Is the default scheduler using commands like Get from the internal client? Ideally we can figure out how FluxSetup can manage this interaction. My worry about using this waiting/running queue as the communication between the FluxJob / FluxSetup is that it could easily get out of sync with what is actually going on. 

> 5/6a. Are the StatefulSet Pods created at this stage as components of the MiniCluster by the FluxJob reconciler? 

Yes - when the job is moved from waiting -> run (meaning it has permission to run) and it's status (Status.Conditions) is waiting, we know it was just moved and should do the steps / reconciles to make the stateful set and configs until we hit the bottom, and then we flag it (via condition again) as ready. A lot of this design was trying to get these two different things to communicate knowing that there isn't supposed to be communication, and they are supposed to make decisions based on the state of the cluster observed. Maybe making the FluxManager and then these indirect communications is cheating, or just bad design, but I'm ok to learn from it and try something better!

> What determines when the resources are available (i.e. K8s default scheduler or KubeFlux)? What criteria allow the job to enter the queue heap (is allowed to run)?

I'm not sure what KubeFlux is (is that another operator?) but right now I have a MaxSize in the FluxSetup spec, and (theoretically) although it doesn't do a check now (since I'm developing everything is allowed to move from waiting to running)  it will compare the running set to the waiting set, and calculate if there is room to add another one. It's not any kind of elegant scheduling - it's a high level "can I make another MiniCluster given what I have."

> 5/6b. "Moving from waiting -> the manager->queue->heap indicates running, and this is done by the scheduler." This is the Flux Operator scheduler, not the K8s cluster scheduler (e.g., the K8s default), correct?

Yes it's my very stupid for loop that continuously moves from waiting -> running, and doesn't do any checks at the moment. I wanted to figure out how to get the job to launch first, and then I could go back to the overall design. I think the laugh / figuring out the hosts is the harder of the two, and there are many designs we could try (this just being one, and likely not the last because I made it just this weekend!)

> 5/6c. Under what conditions does the MiniCluster transition to READY (StatefulSet Pods + ConfigMap + Secrets are READY)?

When the stateful set, two config maps, and the secrets are generated and (TBA) the /etc/hosts is correct to connect the pods, or in simpler terms when we are able to retrieve each of those (and don't have to create one) and reach the bottom of the newMiniCluster function.

> And unrelated to the enumerated sequence, can the current design work with PodGroups?

I don't know what those are, but I'll read about them tonight! :laughing: 

Thanks for the discussion and the opportunity to work on this! I am still figuring out best practices and what is possible, etc, and it's definitely OK if we wind up trying different things. I am learning / learned a _ton_ this weekend and it was amazing (and am down for more!) Have a good night!

**Comment by vsoch** [comment]
Okay so based on:

- https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/kep/42-podgroup-coscheduling/README.md
- https://github.com/kubernetes-sigs/kube-batch/blob/master/pkg/client/listers/scheduling/v1alpha1/podgroup.go
- https://volcano.sh/en/docs/podgroup/

It sounds like this is a common group of pods (perhaps identified with a common uid filter) that has extra attributes for scheduling / batch jobs. And it is / is related to a scheduling operator? I think as long as there is some common selector for the group and the underlying unit (is a pod?) https://github.com/kubernetes-sigs/kube-batch/blob/86b2ba217ad76340221fdc9b4ec16f3209dbff7e/pkg/scheduler/api/job_info.go and we can have the some overlapping features of the stateful set (e.g., volumes, easy creation) I don't see why we couldn't minimally try. I'm new enough to this space / these concepts that it would be helpful to spec out some ideas to work on.

This is also kind of neat because I stumbled on the RESTclient yesterday, and was creating one to try and launch a job (and I got close but hit a permissions error!) https://github.com/kubernetes-sigs/kube-batch/blob/86b2ba217ad76340221fdc9b4ec16f3209dbff7e/pkg/client/clientset/versioned/typed/scheduling/v1alpha2/podgroup.go#L60-L63 Maybe this could help.

**Comment by ArangoGutierrez** [comment]
Let's talk about this on the Wed meeting

**Comment by vsoch** [comment]
okay sounds good! I was like "oh no three days and I can't work on this that's so long!" but then I remembered we are already at Tuesday so tomorrow isn't so far away. :laughing:

**Review by ArangoGutierrez — COMMENTED** [review]
In Kubernetes, things with a state, are much better off preserved in a CRD/CR than in memory. this second design removes the `mini-cluster` CRD, and I don't like that idea. Let's talk about this during the next call

**Comment by vsoch** [comment]
Here was the start of that design: https://github.com/flux-framework/flux-operator/pull/11
